lightning-macros/src/lib.rs

```diff
@@ -74,3 +77,175 @@ pub fn maybe_await(expr: TokenStream) -> TokenStream {
 
 	quoted.into()
 }
+
+fn expect_ident(token: &TokenTree, expected_name: Option<&str>) {
+	if let TokenTree::Ident(id) = &token {
+		if let Some(exp) = expected_name {
+			assert_eq!(id.to_string(), exp, "Expected ident {}, got {:?}", exp, token);
+		}
+	} else {
+		panic!("Expected ident {:?}, got {:?}", expected_name, token);
+	}
+}
+
+fn expect_punct(token: &TokenTree, expected: char) {
+	if let TokenTree::Punct(p) = &token {
+		assert_eq!(p.as_char(), expected, "Expected punctuation {}, got {}", expected, p);
+	} else {
+		panic!("Expected punctuation {}, got {:?}", expected, token);
+	}
+}
+
+/// Scans a match statement for fields which should be skipped
+///
+/// Wraps a `match self {..}` statement and scans the fields in the match patterns (in the form
+/// `ref $field_name: $field_ty`) for types marked `legacy`, skipping those fields.
+#[proc_macro]
+pub fn skip_legacy_fields(expr: TokenStream) -> TokenStream {
```

Comment:
Maybe we can break the function into modular parts. Something like `process_match_pattern` for handling the `Enum::Variant` part and a `process_field` for the `internal fields`. It might help keep things organized and easier to follow!